File: 500/560 - Subarray Sum Equals K.java
import java.util.HashMap;

class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> hm = new HashMap<Integer, Integer>();
        int sum = 0, ret = 0;
        
        hm.put(0, 1);
        for (int num : nums) {
            sum += num;
            
            if (hm.containsKey(sum - k))
                ret += hm.get(sum - k);
            
            if (hm.containsKey(sum))
                hm.replace(sum, hm.get(sum) + 1);
            else
                hm.put(sum, 1);
        }
        
        return ret;
    }
}

File: csharp/560.subarray-sum-equals-k.cs
/*
 * @lc app=leetcode id=560 lang=csharp
 *
 * [560] Subarray Sum Equals K
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int SubarraySum(int[] nums, int k) {
        IDictionary<int, int> count = new Dictionary<int, int>();
        int ret = 0, sum = 0;
        count[0] = 1;
        foreach (int num in nums) {
            sum += num;
            int c = 0;
            count.TryGetValue(sum - k, out c);
            ret += c;
            count.TryGetValue(sum, out c);
            count[sum] = 1 + c;
        }
        return ret;
    }
}
// @lc code=end

